Moving bed with bidirectional spatial discretization

Robbybp opened the pull request:
## Fixes singularities due to incompatible discretization for counter-current flow

## Summary/Motivation:
For counter-current flow systems, spatial discretizations of the two phases should have "opposite direction", e.g. forward for the solid phase and backward for the gas. To approximate finite volume methods, these should be "upwind" with respect to direction of flow. This PR adds options to the moving bed unit model to allow users to specify discretizations of the two phases independently.

## Changes proposed in this PR:
- Add `gas_transformation_scheme` and `solid_transformation_scheme` config options to the moving bed unit model. These are only supported if `dae.finite_difference` is used for `transformation_method`

### Legal Acknowledgement

By contributing to this software project, I agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the license terms described in the LICENSE.txt file at the top level of this directory.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

bpaul4 reviewed (commented):
This looks very nice, and well commented with useful exceptions and error messages.

One suggestion, would it be more convenient for users to set a `bed_configuration=[COCURRENT, COUNTERCURRENT]` argument? This would be more intuitive and if users don't pass discretization flags the script could automatically set the discretization to `gas_transformation_scheme=BACKWARD` and `solid_transformation_scheme=FORWARD` for countercurrent beds.

andrewlee94 commented on idaes/models_extra/gas_solid_contactors/unit_models/moving_bed.py:
```diff
@@ -143,6 +144,40 @@ class MBRData(UnitModelBlockData):
+    CONFIG.declare(
+        "gas_transformation_scheme",
+        ConfigValue(
+            default=None,
```
Seeing as this defaults to `"BACKWARD"` anyway, I would suggest that you just set the default value as such and remove `None` from the domain (that said, `None` is always a valid value for the domain, even if it is not explicitly declared).

Robbybp replied:
This doesn't default to `"BACKWARD"`, it defaults to whatever `transformation_scheme` is. Having `None` as the default is important as it lets us check whether these arguments are set.

andrewlee94 replied:
Ah - I see now. However, the doc string mentions Lagrange-Radau, which is not in the `In()` declaration.

Also, now that there are schemes for each side separately, is the declaration above still required (an overall transformation scheme)?

Robbybp replied:
Good catch. Lagrange-Radau is not supported if gas/solid are set individually. This is caught by the `transformation_method != "dae.finite_difference"` check on line 409, which is tested. I just updated the docstring.

`transformation_scheme` is still supported for backward compatibility, convenience, and to support collocation for the length domain (which may be useful for steady state).

andrewlee94 commented on idaes/models_extra/gas_solid_contactors/unit_models/moving_bed.py:
```diff
@@ -143,6 +144,40 @@ class MBRData(UnitModelBlockData):
+        "solid_transformation_scheme",
+        ConfigValue(
+            default=None,
+            domain=In([None, "BACKWARD", "FORWARD"]),
```
As above.

andrewlee94 commented on idaes/models_extra/gas_solid_contactors/unit_models/moving_bed.py:
```diff
@@ -433,12 +498,49 @@ def build(self):
+            # Create length domain sets for gas and solid phases. This is
+            # necessary so we can apply different discretizations to the
+            # different phases.
+            self._gas_tr_scheme = self.config.gas_transformation_scheme
```
Is it necessary to assign these aliases to the model object, as opposed to just using them locally within the code?

Robbybp replied:
It is necessary as they need to be used in the `_apply_transformation` method

Robbybp replied:
Alternatively, the same thing could be accomplished by augmenting the call signature of `_apply_transformation`.

andrewlee94 commented on idaes/models_extra/gas_solid_contactors/unit_models/moving_bed.py:
```diff
@@ -433,12 +498,49 @@ def build(self):
+            # Arbitrarily, use the solid phase length domain.
+            #
+            # We do not use Reference as it does not work for Sets.
+            super(_BlockData, self).__setattr__(
```
Minor point, but we would normally use `add_object_reference` for this (as this is what it does behind the scenes).

andrewlee94 commented on idaes/models_extra/gas_solid_contactors/unit_models/moving_bed.py:
```diff
@@ -433,12 +498,49 @@ def build(self):
+        if (
+            # solid and gas transformation schemes are specified
+            (self.config.solid_transformation_scheme is not None)
+            and (self.config.gas_transformation_scheme is not None)
```
A minor question, but would it be better to just check that the transformation schemes are different here - i.e. the real point is that if they are different then we need separate domains.

However, going a step further, I would suggest we consider just having two domains at all times (regardless of the transformations used). In the current approach, the model structure is different depending on the choices made, which may have unforeseen consequences, so it might be worth just using two domains at all times to avoid this.

Robbybp replied:
My thought was that, if schemes are provided for solid and gas, they should have different length domains. In this way, whether or not we have two length domains depends on whether these arguments were provided, rather than the value of the arguments. I don't feel strongly about this, however.

I would rather not *always* have two length domains, however, as this would break code that relies on the phases having the same length domain (I have written such code myself). I agree that always having two length domains is a cleaner solution, however.

andrewlee94 commented on idaes/models_extra/gas_solid_contactors/unit_models/tests/test_MB.py:
```diff
@@ -1123,3 +1132,335 @@ def test_initialization_error(self, iron_oc):
+        ):
+            m.fs.unit2.build()
+
+        # with pytest.raises(
```
Commented code.

Robbybp commented:
@bpaul4 
> This looks very nice, and well commented with useful exceptions and error messages.
> 
> One suggestion, would it be more convenient for users to set a `bed_configuration=[COCURRENT, COUNTERCURRENT]` argument? This would be more intuitive and if users don't pass discretization flags the script could automatically set the discretization to `gas_transformation_scheme=BACKWARD` and `solid_transformation_scheme=FORWARD` for countercurrent beds.

There currently is a `flow_type` option that must be set to `"counter_current"`. I think it would make sense to add an option for co-current, however, that is outside the scope of this PR. I also agree that it would be nice to automatically set the discretization schemes based on the "flow type", but would rather not do this here and now as it would be a breaking change. It is also worth noting that, as far as I know/remember, a "backward-backward" discretization (the current default) is fine (at least from a singularity perspective) for a steady state model.

Robbybp commented:
I don't think I can make the dev meeting tomorrow, but this is ready to merge, as far as I'm concerned. Can somebody let me know here if it's being held up for any reason (e.g. the failing tests).

andrewlee94 commented:
I would like @aostace01 to take a look at this as someone with more knowledge of the model, but I just noticed that she does not appear to be tagged on this PR.

aostace01 reviewed (commented):
Robby, this is a great new feature for the moving bed model; thank you for contributing it. Everything looks good to me.
 
There is one thing that comes to mind though: the [documentation](https://github.com/IDAES/idaes-pse/blob/main/docs/reference_guides/model_libraries/gas_solid_contactors/unit_models/moving_bed.rst) - it should reflect this new option of having separate discretization schemes for the gas and solid phases. A few sentences explaining that having just one scheme for both (that could be useful for steady state studies) remains an option would be helpful, too; maybe as a comment in the code, or documentation, or both - not sure what the consensus is on this.

Robbybp commented:
@aostace01 I just updated the config docstrings to clarify the exclusivity and dependence of the `transformation_scheme`, `gas_transformation_scheme`, and `solid_transformation_scheme` options. This will appear in the "MBR Class" section of the docs. 

I also added a small "Discretization" section near the top of the doc page, which recommends using these new options for dynamic operation.

aostace01 reviewed (approved):
Robby, thank you for updating the config docstrings, and for adding a paragraph to the MBR documentation.
